{% extends "base.html" %}
  
{% block title %}About Phage Host Finder{% endblock %}

{% block page_content %}
<div class="page-header"><h1> About Phage Host Finder</h1></div>
<div class="container">
    <div class="row">
        <h4>Background:</h4>
        <div class="col-md-11">
            <div>
                <p>
                    Antibiotic resistance is disseminating into a global crisis. Bacteria are developing immunity to most everyday drugs.
                    Overprescribing such drugs, along with a lack of development of new drugs, has set the stage where superbugs are now a commonplace occurrence.
                    Consequently, a significant amount of research is being redirected to tackle this rising problem.
                </p>
            </div>
            <br/>
            <div>
                <p>
                    Nature offers a natural response to bacteria – Bacteriophages (phages). Phages, aka bacteria-eaters, infect and destroy bacteria cells.
                    But there is one caveat. These phages are very specific to hosts. Hence, host ranges are not easily identifiable.
                    Multiple avenues of explorations, such as viral tagging, wet lab research, and computational methods have been undertaken to assist with phage-host identification.
                    Phage Host Finder aims to assist in the research by providing a resource to predict phage-host interactions more efficiently.
                    This site focuses on a unique feature, the attP/attB sites, a homologous sequence present in both temperate phage and bacteria respectively.
                    These sites are the attachment sites where the phage integrates its DNA into the bacteria genome and are frequently found adjacent to the integrase gene in the phage genome and within or near a tRNA gene in the bacterial genome.
                    A match between phage and bacteria would indicate the likelihood of a successful phage infection of that host.
                </p>
            </div>
            <br/>
            <div>
                <p>
                    The integrase gene is a critical aspect of this search. Not all phages are temperate, meaning that they don’t possess the ability to integrate themselves into the genome of bacteria. These phages are labeled as lytic phages, and this method cannot be used to identify their hosts.
                    The temperate phages are referred to as lysogenic phages, and possess either of two integrase genes; serine or tyrosine.
                </p>
            </div>
        </div>
    </div>

    <div class="row">
        <h4>Input:</h4>
        <div class="col-md-11">
            <div>
                <p>
                    Phage Host finder currently accepts only accession numbers of annotated sequences as inputs. The input sequence must provide the CDS feature and the "/product" tag.
                    Future updates will include additional options.
                </p>
            </div>
        </div>
        <div class="col-md-6">
            <br/>
            <div><b> Example of CDS Features: </b></div>
            <br/>
            <div>
                <img src="../static/images/CDS_feature.png" class="img-responsive" alt="CDS Feature">
            </div>
            <br/>
        </div>
        <div class="col-md-6">
            <br>
            <div><b> Example of Accession# Z18946:</b></div>
            <br/>
            <div><img src="../static/images/screenshot1.png" class="img-responsive" alt="Accession Example Input"></div>
            <br/>
        </div>
        <div class="col-md-11">
            <br/>
            <p class="text-danger"><b>**Please note that the search for the attP/attB homology region can take some time.
                Please be patient while the process is ongoing and do not refresh the page.</b></p>
        </div>
    </div>

    <div class="row">
        <h4>Output:</h4>
        <div class="col-md-11">
            <p>
            The e-value set for the search is <0.01. Virtually identical short alignments have relatively high E values.
            This is because the calculation of the E value takes into account the length of the query sequence.
            These high E values make sense because shorter sequences have a higher probability of occurring in the database purely by chance.
            (<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=FAQ#expect">NCBI BLAST</a>)
            </p>
            <br/>
        </div>
        <div class="col-md-11">
            <p>The results will return a table listing the families in the Actinobacteria phylum that the attP/attB sequence was located in, along with the total number of hits within the respective families.
                The more hits within a family, the stronger the likelihood of a familial host match. The results will also include a consensus sequence with core nucleotides highlighted.
                The number of core nucleotides will vary extensively, depending on the type of integrase gene.
                A serine integrase gene will return a core of approximately 2-12 base pairs while a tyrosine integrase gene will return approximately 20-45 core nucleotides.
            </p>
            <br/>
        </div>
        <div class="col-md-6">
            <div><b> Results for Accession# Z18946:</b></div>
            <img src="../static/images/results1.png" class="img-responsive" alt="Accession Example Output">
        </div>
        <div class="col-md-11">
            <br/>
            <p>If no results are returned, it can be attributed to a few possibilities. This site is currently in its first phase of development and is limited in scope.
                If the phage hasn't been annotated yet or the attP/attB recognition sequence lies further away from the integrase gene, this site currently does not have the capabilities of recognizing those variations.
                Another likely possibility to consider is that the phage may not be temperate one, and therefore doesn't possess this mechanism of integration.
            </p>
        </div>
    </div>
</div>

<div class="container">
    <div class="row">
        <h4>Future Site updates</h4>
        <div>
            <ul>
                <li> Accept multiple input formats </li>
                <li> Allow user to set e-value </li>
                <li> Expand database to others outside of the Actinobacteria phylum</li>
            </ul>
        </div>
    </div>
</div>

{% endblock %}
~                
